Mathlib/Util/CompileInductive.lean

```diff
@@ -0,0 +1,196 @@
+/-
+Copyright (c) 2023 Parth Shastri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Parth Shastri
+-/
+import Lean.Compiler.CSimpAttr
+import Lean.Elab.PreDefinition
+
+/-!
+# Define the `#compile inductive` command.
+
+The command `#compile inductive Foo` adds compiled code for the recursor `Foo.rec`,
+working around a bug in the core Lean compiler which does not support recursors.
+
+Similarly, `#compile def Foo.foo` adds compiled code for definitions when missing.
+This can be the case for type class projections, or definitions like `List._sizeOf_1`.
+-/
+
+namespace Mathlib.Util
+
+open Lean
+
+private def replaceConst (old new : Name) (e : Expr) : Expr :=
+  e.replace λ | .const n us => if n == old then some (.const new us) else none | _ => none
+
+open Meta
+
+private def mkFunExts' (xs : Array Expr) (e : Expr) (βfg : Expr × Expr × Expr) : MetaM Expr := do
+  let mut (β, f, g) := βfg
+  let mut e := e
+  for x in xs.reverse do
+    let α ← inferType x
+    f ← mkLambdaFVars #[x] f
+    g ← mkLambdaFVars #[x] g
+    e := mkApp5 (.const ``funext [(← inferType α).sortLevel!, (← inferType β).sortLevel!])
+      α (← mkLambdaFVars #[x] β) f g (← mkLambdaFVars #[x] e)
+    β ← mkForallFVars #[x] β
+  return e
+
+private def mkFunExts (e : Expr) : MetaM Expr := do
+  forallTelescope (← inferType e) λ xs body => do
+    let some βfg := (← whnf body).eq?
+      | throwError "expected equality"
+    mkFunExts' xs (mkAppN e xs) βfg
+
+private def mkEq (α a b : Expr) : MetaM Expr := do
+  return mkApp3 (.const ``Eq [(← inferType α).sortLevel!]) α a b
+
+private def mkEqRefl (α a : Expr) : MetaM Expr := do
+  return mkApp2 (.const ``Eq.refl [(← inferType α).sortLevel!]) α a
+
+open Elab
+
+/--
+Compile the definition `dv` by adding a second definition `dv✝` with the same body,
+and registering a `csimp`-lemma `dv = dv✝`.
+-/
+def compileDefn (dv : DefinitionVal) : TermElabM Unit := do
+  let name ← mkFreshUserName dv.name
+  addAndCompile <| .defnDecl { dv with name }
+  let levels := dv.levelParams.map .param
+  let old := .const dv.name levels
+  let new := .const name levels
+  let name ← mkFreshUserName <| dv.name.str "eq"
+  addDecl <| .thmDecl {
+    name
+    levelParams := dv.levelParams
+    type := ← mkEq dv.type old new
+    value := ← mkEqRefl dv.type old
+  }
+  Compiler.CSimp.add name .global
+
+/--
+`#compile def Foo.foo` adds compiled code for the definition `Foo.foo`.
+This can be used for type class projections or definitions like `List._sizeOf_1`,
+for which Lean does not generate compiled code by default
+(since it is not used 99% of the time).
+-/
+elab tk:"#compile " "def " i:ident : command => Command.liftTermElabM do
```

Comment:
This shouldn't have a `#` on the name, because it is not a temporary use command. Rather, I think it should have a `%` suffix similar to `gen_injective_theorems%` to reflect that it is somewhat internal-use and also a workaround for a current lean limitation.

Reply:
It's `compile_def%` and `compile_inductive%` now.